File: Java_Solutions/LongestSubstringWithoutRepChars.java
class Solution {
    //Given a string s, find the length of the longest substring without repeating characters.
    public int lengthOfLongestSubstring(String s) {
        Set<Character> visitedChars = new HashSet<Character>();
        int res = 0;
        int l = 0;
        int n = s.length();

        for(int r=0; r<n; r++){
            while(visitedChars.contains(s.charAt(r))){
                visitedChars.remove(s.charAt(l));
                l+=1;
            }
            visitedChars.add(s.charAt(r));
            res = Math.max(res, r-l+1);
        }
        return res;
    }
}

File: Python_Solutions/LongestSubstringWithoutRepChars.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen_chars   = set() #set of characters seen already in the string
        max_length   = 0     #length of longest substring without repeating chars
        l = 0
        n = len(s)
        for r in range(n):
            while s[r] in seen_chars:
                seen_chars.remove(s[l])
                l+=1
            seen_chars.add(s[r])
            max_length = max(max_length, r-l+1)

        return max_length
